File: Cost-effectiveness/Decision Trees/Extended Tree - XDR-TB/Code/Main/Optimism correction/AverageNMB_CI_Investigation.py
import numpy as np
import pandas as pd
from scipy.stats import beta, norm, t
import matplotlib.pyplot as plt
import seaborn as sns

# Set seed for reproducibility
np.random.seed(42)

# Fixed parameters
wtp = 30000
n_patients = 540
nsim_values = [10, 50, 100, 500, 1000, 2000, 5000]

# Fixed patient illness status
p_ill = 0.19
ill_status = np.random.binomial(1, p_ill, size=n_patients)

p_cure_ill_treat = 0.85
p_cure_notill = 0.99

# ========= Average NMB by patient
results_pt = []

for nsim in nsim_values:
    mean_nmbs = []

    for j in range(n_patients):
        cost_cure_ill = norm.rvs(loc=20000, scale=200, size=nsim)
        cost_death_ill = norm.rvs(loc=100000, scale=200, size=nsim)
        cost_cure_notill = norm.rvs(loc=30000, scale=200, size=nsim)
        cost_death_notill = norm.rvs(loc=100000, scale=200, size=nsim)
        qaly_cure_ill = beta.rvs(a=90, b=10, size=nsim)
        qaly_cure_notill = beta.rvs(a=50, b=50, size=nsim)

        ill_status_pt = np.repeat(ill_status[j], nsim)

        qaly = np.where(
            ill_status_pt == 1,
            p_cure_ill_treat * qaly_cure_ill,
            p_cure_notill * qaly_cure_notill
        )
        cost = np.where(
            ill_status_pt == 1,
            p_cure_ill_treat * cost_cure_ill + (1 - p_cure_ill_treat) * cost_death_ill,
            p_cure_notill * cost_cure_notill + (1 - p_cure_notill) * cost_death_notill
        )

        nmb = qaly * wtp - cost
        mean_nmbs.append(np.mean(nmb))

    mean_nmb = np.mean(mean_nmbs)
    se_nmb = np.std(mean_nmbs, ddof=1) / np.sqrt(n_patients)
    t_crit = t.ppf(0.975, df=n_patients - 1)
    ci_lower = mean_nmb - t_crit * se_nmb
    ci_upper = mean_nmb + t_crit * se_nmb

    results_pt.append({
        "nsim": nsim,
        "mean_nmb": mean_nmb,
        "ci_lower": ci_lower,
        "ci_upper": ci_upper
    })

df_pt = pd.DataFrame(results_pt)

# Plot
sns.set(style="whitegrid")
plt.figure(figsize=(10, 6))
plt.plot(df_pt["nsim"], df_pt["mean_nmb"], marker='o', color='steelblue')
plt.fill_between(df_pt["nsim"], df_pt["ci_lower"], df_pt["ci_upper"], color='steelblue', alpha=0.2)
plt.ylim(-15000, -14000)
plt.title("Expected NMB by patient with 95% CI vs. Number of Parameter Simulations (Per Patient)")
plt.xlabel("Number of Simulations (nsim)")
plt.ylabel("Mean Net Monetary Benefit (Treat All Patients)")
plt.show()

# ========= Average NMB by simulation
results_sim = []

for nsim in nsim_values:
    mean_nmbs = []

    for k in range(nsim):
        cost_cure_ill = norm.rvs(loc=20000, scale=200)
        cost_death_ill = norm.rvs(loc=100000, scale=200)
        cost_cure_notill = norm.rvs(loc=30000, scale=200)
        cost_death_notill = norm.rvs(loc=100000, scale=200)
        qaly_cure_ill = beta.rvs(a=90, b=10)
        qaly_cure_notill = beta.rvs(a=50, b=50)

        qaly = np.where(
            ill_status == 1,
            p_cure_ill_treat * qaly_cure_ill,
            p_cure_notill * qaly_cure_notill
        )
        cost = np.where(
            ill_status == 1,
            p_cure_ill_treat * cost_cure_ill + (1 - p_cure_ill_treat) * cost_death_ill,
            p_cure_notill * cost_cure_notill + (1 - p_cure_notill) * cost_death_notill
        )

        nmb = qaly * wtp - cost
        mean_nmbs.append(np.mean(nmb))

    mean_nmb = np.mean(mean_nmbs)
    se_nmb = np.std(mean_nmbs, ddof=1) / np.sqrt(nsim)
    t_crit = t.ppf(0.975, df=nsim - 1)
    ci_lower = mean_nmb - t_crit * se_nmb
    ci_upper = mean_nmb + t_crit * se_nmb

    results_sim.append({
        "nsim": nsim,
        "mean_nmb": mean_nmb,
        "ci_lower": ci_lower,
        "ci_upper": ci_upper
    })

df_sim = pd.DataFrame(results_sim)

# Plot
plt.figure(figsize=(10, 6))
plt.plot(df_sim["nsim"], df_sim["mean_nmb"], marker='o', color='steelblue')
plt.fill_between(df_sim["nsim"], df_sim["ci_lower"], df_sim["ci_upper"], color='steelblue', alpha=0.2)
plt.ylim(-15000, -14000)
plt.title("Expected NMB by simulation with 95% CI vs. Number of Parameter Simulations (Per Simulation)")
plt.xlabel("Number of Simulations (nsim)")
plt.ylabel("Mean Net Monetary Benefit (Treat All Patients)")
plt.show()

File: Cost-effectiveness/Decision Trees/Extended Tree - XDR-TB/Analysis/AverageNMB_CI_investigation.R
set.seed(42)
library(dplyr)
library(ggplot2)

# Fixed parameters
wtp <- 30000
n_patients <- 540  # fixed number of patients

# Simulation sizes to evaluate
nsim_values <- c(10, 50, 100, 500, 1000, 2000, 5000)

# Fixed patient illness status
p_ill <- 0.19
ill_status <- rbinom(n_patients, 1, p_ill)

p_cure_ill_treat <- 0.85
p_cure_notill <- 0.99

#========= Average NMB by patient

# Store results
results_pt <- data.frame()

for (nsim in nsim_values) {

  mean_nmbs = c()
  
  for(j in 1:n_patients) {
    # Store mean NMBs from each parameter simulation
    
    cost_cure_ill <- rnorm(nsim, mean = 20000, sd = 200)
    cost_death_ill <- rnorm(nsim, mean = 100000, sd = 200)
    cost_cure_notill <- rnorm(nsim, mean = 30000, sd = 200)
    cost_death_notill <- rnorm(nsim, mean = 100000, sd = 200)
    qaly_cure_ill <- rbeta(nsim, 90, 10)
    qaly_cure_notill <- rbeta(nsim, 50, 50)
    
    ill_status_pt = rep(ill_status[j], nsim)
      
    qaly <- ifelse(ill_status_pt == 1, p_cure_ill_treat * qaly_cure_ill , qaly_cure_notill * p_cure_notill)
    cost <- ifelse(ill_status_pt == 1, p_cure_ill_treat * cost_cure_ill + (1-p_cure_ill_treat)*cost_death_ill, 
                     p_cure_notill * cost_cure_notill + (1-p_cure_notill)*cost_death_notill)
      
    nmb <- qaly * wtp - cost
    mean_nmbs[j] <- mean(nmb)
  }
  
  # Calculate mean and t-distribution CI
  mean_nmb <- mean(mean_nmbs)
  se_nmb <- sd(mean_nmbs) / sqrt(n_patients)
  t_crit <- qt(0.975, df = n_patients - 1)
  ci_lower <- mean_nmb - t_crit * se_nmb
  ci_upper <- mean_nmb + t_crit * se_nmb
  
  results_pt <- rbind(results_pt, data.frame(
    nsim = nsim,
    mean_nmb = mean_nmb,
    ci_lower = ci_lower,
    ci_upper = ci_upper
  ))
}

# Plotting the results_pt
ggplot(results_pt, aes(x = nsim, y = mean_nmb)) +
  geom_line(color = "steelblue") +
  geom_point(color = "steelblue") +
  geom_ribbon(aes(ymin = ci_lower, ymax = ci_upper), alpha = 0.2, fill = "steelblue") +
  coord_cartesian(ylim = c(-15000, -14000)) +
  theme_minimal() +
  labs(
    title = "Expected NMB with 95% t-based CI vs. Number of Parameter Simulations",
    x = "Number of Simulations (nsim)",
    y = "Mean Net Monetary Benefit (Treat All Patients)"
  )


#========= Average NMB by simulation

# Store results
results_sim <- data.frame()

for (nsim in nsim_values) {
  
  mean_nmbs = c()

  for(k in 1:nsim) {
      # Store mean NMBs from each parameter simulation
      
    cost_cure_ill <- rnorm(1, mean = 20000, sd = 200)
    cost_death_ill <- rnorm(1, mean = 100000, sd = 200)
    cost_cure_notill <- rnorm(1, mean = 30000, sd = 200)
    cost_death_notill <- rnorm(1, mean = 100000, sd = 200)
    qaly_cure_ill <- rbeta(1, 90, 10)
    qaly_cure_notill <- rbeta(1, 50, 50)
    
    qaly <- ifelse(ill_status == 1, p_cure_ill_treat * qaly_cure_ill , qaly_cure_notill * p_cure_notill)
    cost <- ifelse(ill_status == 1, p_cure_ill_treat * cost_cure_ill + (1-p_cure_ill_treat)*cost_death_ill, 
                   p_cure_notill * cost_cure_notill + (1-p_cure_notill)*cost_death_notill)
    
    nmb <- qaly * wtp - cost
    mean_nmbs[k] <- mean(nmb)
  } 
  # Calculate mean and t-distribution CI
  mean_nmb <- mean(mean_nmbs)
  se_nmb <- sd(mean_nmbs) / sqrt(nsim)
  t_crit <- qt(0.975, df = nsim - 1)
  ci_lower <- mean_nmb - t_crit * se_nmb
  ci_upper <- mean_nmb + t_crit * se_nmb
  
  results_sim <- rbind(results_sim, data.frame(
    nsim = nsim,
    mean_nmb = mean_nmb,
    ci_lower = ci_lower,
    ci_upper = ci_upper
  ))
}



# Plotting the results
ggplot(results_sim, aes(x = nsim, y = mean_nmb)) +
  geom_line(color = "steelblue") +
  geom_point(color = "steelblue") +
  geom_ribbon(aes(ymin = ci_lower, ymax = ci_upper), alpha = 0.2, fill = "steelblue") +
  coord_cartesian(ylim = c(-15000, -14000)) +
  theme_minimal() +
  labs(
    title = "Expected NMB with 95% t-based CI vs. Number of Parameter Simulations",
    x = "Number of Simulations (nsim)",
    y = "Mean Net Monetary Benefit (Treat All Patients)"
  )
